lib/timeline_records.rb

```diff
@@ -0,0 +1,298 @@
+        fields = details[event_type]
+        records = @patient.send(event_type)
+        records = Array(records)
+
+        records.each do |record|
+          event_details =
+            fields.each_with_object({}) do |field, hash|
+              field_value = record.send(field)
+              hash[field.to_s] = field_value.nil? ? "nil" : field_value
+            end
+          @events << {
+            event_type: record.class.name,
+            id: record.id,
+            details: event_details,
+            created_at: record.created_at
+          }
+        end
+      else
+        custom_event_handler(event_type)
+      end
+    end
+    @events.sort_by! { |event| event[:created_at] }.reverse!
+  end
+
+  def load_grouped_events(event_names)
+    load_events(event_names)
+    @events.each { |event| event[:details] = format_details(event) }
+    @grouped_events_by_date =
+      @events.group_by do |event|
+        event[:created_at].strftime(Date::DATE_FORMATS[:long])
+      end
+    @grouped_events =
+      @grouped_events_by_date.sort_by { |date, _events| date }.reverse!.to_h
+  end
+
+  private
+
+  def format_details(event)
+    if event[:details].is_a?(Hash)
+      event[:details].map { |k, v| "#{k}: #{v}" }.join(", ")
+    else
+      event[:details].to_s
+    end
+  end
+
+  def custom_event_handler(event_type)
+    case event_type
+    when :org_cohort_imports
+      @events += org_cohort_imports_events
+    when /^add_class_imports_\d+$/ # e.g. add_class_imports_123
+      session_id = event_type.to_s.split("_").last
+      @events += add_class_imports_events(session_id)
+    when :audits
+      @events += audits_events
+    else
+      puts "No handler for event type: #{event_type}"
+    end
+  end
+
+  def org_cohort_imports_events
+    @additional_events[:cohort_imports].map do |cohort_import_id|
+      cohort_import = CohortImport.find(cohort_import_id)
+      {
+        event_type: "CohortImport",
+        id: cohort_import.id,
+        details: "excluding patient",
+        created_at: cohort_import.created_at
+      }
+    end
+  end
+
+  def add_class_imports_events(session_id)
+    @additional_events[:class_imports][session_id.to_i].map do |class_import_id|
+      class_import = ClassImport.find(class_import_id)
+      {
+        event_type: "ClassImport",
+        id: class_import.id,
+        details: {
+          session_id: "#{class_import.session_id}, excluding patient"
+        },
+        created_at: class_import.created_at.to_time
+      }
+    end
+  end
+
+  def audits_events
+    audit_events =
+      (
+        if audits[:include_associated_audits]
+          @patient.own_and_associated_audits
+        else
+          @patient.audits
+        end
+      )
+
+    audit_events.map do |audit|
+      filtered_changes =
+        audit
+          .audited_changes
+          .transform_keys(&:to_s)
+          .each_with_object({}) do |(key, value), hash|
+            if ALLOWED_AUDITED_CHANGES.include?(key.to_sym)
```

Comment:
Here we're transforming the keys to strings (`to_s`), and then later on back to symbols again (`to_sym`), is there a way we can avoid this extra processing?

I'm wondering if `audited_changes` will already have string keys since it'll be coming from the database.